### PR TITLE
CXX: Add some new fields to the cxx parser

### PR DESCRIPTION
**Pull request by pragmaware**
- Add template parameter support in class/struct/union declarations and function prototypes/definitions.
- Add "end" field for marking end lines of classes, unions, structs, namespaces, enums function declarations and lambdas
- Add support for extracting lambda capture lists

### CODE REVIEW

File: Units/parser-c.r/bit_field.c.d/args.ctags [hunk 1]
```diff
@@ -1 +1,2 @@
 --sort=no
+--fields=+{c.end}
```
Comment (masatake) [inline Units/parser-c.r/bit_field.c.d/args.ctags]: Could you try to use wildcard notation `{end}` here? I hope this notation works well.
`{end}` means `{*.end}`. And `*` means "any language".

I implemented but I don't have a test case for the wildcard notation.
I would like to reuse this as a test case for the notation.

**Comment by masatake** [comment]
13677b4 looks good.
I would like to take time for reviewing  ee27cda.
`end' is very impressive.

File: Units/parser-c.r/bit_field.c.d/args.ctags (hunk 1 above)
Comment (pragmaware) [inline Units/parser-c.r/bit_field.c.d/args.ctags]: Seems to work :)

File: parsers/cxx/cxx_tag.c [hunk 2]
```diff
@@ -189,14 +221,44 @@ void cxxTagSetCPPField(CXXTagCPPField eField,const char * szValue)
...
+
+int cxxTagCommit(void)
 {
 	if(g_oCXXTag.isFileScope)
 	{
 		if (isXtagEnabled(XTAG_FILE_SCOPE))
 			markTagExtraBit (&g_oCXXTag, XTAG_FILE_SCOPE);
 		else
-			return;
+			return SCOPE_NIL; // FIXME: why the "invalid" cork queue index is SCOPE_NIL?
```
Comment (masatake) [inline parsers/cxx/cxx_tag.c]: I agree. CORK_NIL is better?

**Comment by masatake** [comment]
LGTM. make units LANGUAGE=C++ VG=1 is passed.

**Comment by masatake** [comment]
Please, wait for merging. I found a critical bug in my code(main part).

**Comment by masatake** [comment]
You understand corkAPI very well.

cxxTagSetCorkQueueCField and cxxTagSetCorkQueueCPPField are needed.
I think this should be part of main side.

Please, merge the commits in this PR.
I will fix main part later.

**Comment by masatake** [comment]
What I should provide is
```
extern int attachFieldToCorkEntry (fieldType type, int index,
				   const char *value)
{
	tagEntryInfo * tag = getEntryInCorkQueue(index);

	if (index == SCOPE_NIL)
		return -1;

	return attachField (type, tag, eStrdup (value);
}
```

**Comment by masatake** [comment]
Macros can be handled well based on your work.
```
[yamato@x201]~/var/ctags-github% cat /tmp/foo.c 
cat /tmp/foo.c 
#define F(X) X
#define G
#include <S>
#define Z(Y)					\
int main (void)					\
{						\
  return 0;					\
}
[yamato@x201]~/var/ctags-github%  u-ctags --fields='+{end}{line}' -o - /tmp/foo.c 
 u-ctags --fields='+{end}{line}' -o - /tmp/foo.c 
F	/tmp/foo.c	/^#define F(/;"	d	line:1	file:	end:1
G	/tmp/foo.c	/^#define G$/;"	d	line:2	file:	end:2
Z	/tmp/foo.c	/^#define Z(/;"	d	line:4	file:	end:8
```

**Comment by masatake** [comment]
Let's merge them!

**Pull request merged**